sources/events.go

```
@@ -0,0 +1,131 @@
+type EventsSource interface {
+	// Returns all new events since GetEvents was last called.
+	GetEvents() ([]kubeapi.Event, error)
+}
+
+// NewEventsSource initializes a new events source and starts a
+// goroutine to watch/fetch event updates.
+func NewEventsSource(client *kubeclient.Client) EventsSource {
+	eventsChan := make(chan eventsUpdate)
+	glog.V(4).Infof("Starting event source")
+	go watchLoop(client.Events(kubeapi.NamespaceAll), eventsChan)
+	glog.V(4).Infof("Finished starting event source")
+	return &eventsSourceImpl{client, eventsChan}
+}
+
+// eventsUpdate is the wrapper object used to pass new events around
+type eventsUpdate struct {
+	events *kubeapi.EventList
+}
+
+// eventsSourceImpl is an implmentation of eventsSource
+type eventsSourceImpl struct {
+	*kubeclient.Client
+	eventsChannel <-chan eventsUpdate
+}
+
+// GetEvents returns all new events since GetEvents was last called.
+func (eventSource *eventsSourceImpl) GetEvents() ([]kubeapi.Event, error) {
+	events := []kubeapi.Event{}
+UpdateLoop:
+	for {
+		select {
+		case eventsUpdate := <-eventSource.eventsChannel:
+			if eventsUpdate.events == nil {
+				return nil, errors.New("Error: recieved a nil event list.")
+			}
+			if eventsUpdate.events.Items == nil {
+				return nil, errors.New("Error: received an event list with nil Items.")
+			}
+			for _, event := range eventsUpdate.events.Items {
+				glog.V(3).Infof("Received new event: %#v\r\n", event)
+				events = append(events, event)
+			}
+		default:
+			break UpdateLoop
+		}
+	}
+	return events, nil
+}
+
+// watchLoop loops forever looking for new events.  If an error occurs it will close the channel and return.
+func watchLoop(eventClient kubeclient.EventInterface, eventsChan chan<- eventsUpdate) {
+	defer close(eventsChan)
+	events, err := eventClient.List(kubelabels.Everything(), kubelabels.Everything())
+	if err != nil {
+		glog.Errorf("Failed to load events: %v", err)
```

In case of an error, we should bubble up the error and try again.